dev/com.ibm.ws.concurrent_fat_policy/test-applications/concurrentpolicyfat/src/web/ConcurrentPolicyFATServlet.java

```diff
@@ -336,4 +337,71 @@ public void testInvokeAllListenersWithLookups() throws Exception {
         assertTrue("unexpected: " + listener.resultOfLookup[STARTING], listener.resultOfLookup[STARTING] instanceof UserTransaction);
         assertTrue("unexpected: " + listener.resultOfLookup[DONE], listener.resultOfLookup[DONE] instanceof UserTransaction);
     }
+
+    /**
+     * When runIfQueueFull is true, it should be possible to submit maxConcurrency number of tasks, even if the maxQueueSize is less
+     * than maxConcurrency because tasks that cannot be enqueued will run on the submitter's thread.
+     */
+    @Test
+    public void testRunWhenQueueFull() throws Exception {
+        CountingTask[] tasks = new CountingTask[4]; // maxConcurrency
+        for (int i = 0; i < 4; i++) {
+            TaskListener listener = i == 0 ? null : new TaskListener().doLookup("java:module/env/concurrent/executor2ref", STARTING, DONE);
+            tasks[i] = new CountingTask(null, listener, null, null);
+        }
+
+        // Occupy the global executor with some other work to increase the chance that tasks submitted to defaultScheduledExecutor
+        // will be unable to queue and run on the submitter's thread.
+        CountingTask blockerTask1 = new CountingTask(null, null, new CountDownLatch(1), new CountDownLatch(1));
+        CountingTask blockerTask2 = new CountingTask(null, null, new CountDownLatch(1), new CountDownLatch(1));
+        Future<Integer> blockerTask1Future = scheduledExecutor2.submit(blockerTask1);
+        cancelAfterTest.add(blockerTask1Future);
+        Future<Integer> blockerTask2Future = scheduledExecutor2.submit(blockerTask2);
+        cancelAfterTest.add(blockerTask2Future);
+
+        // Submit as many tasks as we have maxConcurrency, even though maxQueueSize is only 1
+        List<Future<Integer>> futures = new ArrayList<Future<Integer>>(4);
+        for (int i = 0; i < 4; i++) {
+            Future<Integer> future = defaultScheduledExecutor.submit(tasks[i]);
+            cancelAfterTest.add(future);
+            futures.add(future);
+        }
+
+        // Any tasks that ran on the current thread should already report being done.
+        long curThreadId = Thread.currentThread().getId();
+        if (((TaskListener) tasks[1].listener).threadId[STARTING] == curThreadId)
+            assertTrue(futures.get(1).isDone());
+        if (((TaskListener) tasks[2].listener).threadId[STARTING] == curThreadId)
+            assertTrue(futures.get(2).isDone());
+        if (((TaskListener) tasks[3].listener).threadId[STARTING] == curThreadId)
+            assertTrue(futures.get(3).isDone());
+
+        TaskListener listener;
+        Future<Integer> future = futures.get(0);
+        assertTrue(future.isDone());
```

Comment:
Should you do a future.get() before this to ensure that the task has had time to complete?

Reply:
Yes, the asserts are out of order in that test.  I have attached a commit with the fix.  Thanks for spotting it!